source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -21,76 +23,129 @@ public class MessageExchangeStream : IMessageExchangeStream
 
         readonly Stream stream;
         readonly ILog log;
-        readonly StreamWriter streamWriter;
+        [Obsolete]
+        readonly StreamWriter streamWriter = null;
         readonly IMessageSerializer serializer;
-        readonly Version currentVersion = new Version(1, 0);
-        readonly ControlMessageReader controlMessageReader = new ControlMessageReader();
+        readonly Version currentVersion = new(1, 0);
+        readonly ControlMessageReader controlMessageReader = new();
 
-        public MessageExchangeStream(Stream stream, IMessageSerializer serializer, ILog log)
+        public MessageExchangeStream(Stream stream, IMessageSerializer serializer, AsyncHalibutFeature asyncHalibutFeature, ILog log)
         {
-            #if NETFRAMEWORK
-            this.stream = stream;
-            #else
-            this.stream = new RewindableBufferStream(stream, HalibutLimits.RewindableBufferStreamSize);
-            #endif
+#if NETFRAMEWORK
+            this.stream = asyncHalibutFeature.IsEnabled() ? new TimeoutStream(stream) : stream;
+#else
+            this.stream = asyncHalibutFeature.IsEnabled() ? new RewindableBufferStream(new TimeoutStream(stream), HalibutLimits.RewindableBufferStreamSize) : new RewindableBufferStream(stream, HalibutLimits.RewindableBufferStreamSize);
+#endif
             this.log = log;
-            streamWriter = new StreamWriter(this.stream, new UTF8Encoding(false)) { NewLine = "\r\n" };
             this.serializer = serializer;
+
+#pragma warning disable CS0612 // Type or member is obsolete
+            streamWriter = new StreamWriter(this.stream, new UTF8Encoding(false)) { NewLine = "\r\n" };
```

Comment:
StreamWriter does not support Cancellation on net48. It is kept for sync Halibut but unused for Async Halibut so marked as Obsolete

Reply:
Additionally this is the best path since we don't want to have wrapping streams we don't always use, since that has lead to hard to find bugs.